web-ui/src/test/java/com/tutrit/webclient/controller/EngineerControllerTest.java

```diff
@@ -0,0 +1,119 @@
+import org.springframework.test.web.servlet.MvcResult;
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
+import org.springframework.test.web.servlet.result.MockMvcResultHandlers;
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.mockito.Mockito.when;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+@AutoConfigureMockMvc
+@SpringBootTest(classes = SpringContextConfig.SpringConfig.class)
+class EngineerControllerTest {
+
+    @MockBean
+    EngineerGateway engineerGateway;
+    @Autowired
+    MockMvc mockMvc;
+    @Autowired
+    ObjectMapper objectMapper;
+
+    @Test
+    void findEngineerById() throws Exception {
+        when(engineerGateway.findEngineerById("id1")).thenReturn(Optional.of(createEngineer()));
+
+        final MvcResult result = mockMvc.perform(get("/engineers/id1"))
+                .andExpect(status().isOk())
+                .andExpect(model().attribute("engineer", Matchers.equalTo(createEngineer())))
+                .andExpect(view().name("engineer-form"))
+                .andReturn();
+
+        Engineer actualEngineer = (Engineer) Objects.requireNonNull(result.getModelAndView()).getModel().get("engineer");
+        assertEquals(createEngineer(), actualEngineer);
+    }
+
+    @Test
+    void findUserByIdIfUserNull() throws Exception {
+        when(engineerGateway.findEngineerById("222"))
+                .thenReturn(Optional.empty());
+        MvcResult mvcResult = mockMvc.perform(MockMvcRequestBuilders.get("/engineers/222"))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andExpect(model().attribute("engineer", Matchers.equalTo(engineerHasNullOfFields())))
+                .andExpect(model().attribute("error_404", Matchers.equalTo("Engineer not found")))
+                .andExpect(view().name("engineer-form"))
+                .andReturn();
+        Engineer actualUser = (Engineer) Objects.requireNonNull(mvcResult.getModelAndView()).getModel().get("engineer");
+        assertEquals(engineerHasNullOfFields(), actualUser);
+    }
+
+    @Test
+    void saveEngineer() throws Exception {
+        final MvcResult mvcResult = mockMvc
+                .perform(MockMvcRequestBuilders.post("/engineers/52?" +
+                                "engineerId=52" +
+                                "&firstName=firstName1" +
+                                "&lastName=lastName1" +
+                                "&function=function1" +
+                                "&category=category1" +
+                                "&education=education1" +
+                                "&experience=3" +
+                                "&generalExperience=10")
```

Comment:
You can put all keys and values as parameters as in deleteEngineer() test

Reply:
done